onyo/lib/git.py

```diff
@@ -401,3 +404,85 @@ def check_ignore(self, ignore: Path, paths: list[Path]) -> list[Path]:
+        """Generate a dict from the output of ``git log`` for one commit.
+
+        Internal helper that gets a list of git-log-output lines,
+        that represent a single commit and generates a dictionary from it.
+        """
+        import datetime
+        import re
+        regex_author = re.compile(r"Author:\s+(?P<name>\b.*\b)\s+<(?P<email>[^\s]+)>$")
+        regex_committer = re.compile(r"Commit:\s+(?P<name>\b.*\b)\s+<(?P<email>[^\s]+)>$")
+        commit = dict()
+        for line in lines:
+            if line.startswith('commit '):
+                commit['commit'] = line.split()[1]
+                continue
+            elif line.startswith('Author:'):
+                try:
+                    commit['author'] = re.match(regex_author, line).groupdict()  # pyre-ignore [16] AttributeError is caught
+                except AttributeError as e:
+                    if str(e).endswith("'groupdict'"):
+                        raise RuntimeError(f"Unable to parse author-line:\n{line}") from e
+                    raise
+                continue
+            elif line.startswith('AuthorDate:'):
+                continue
+            elif line.startswith('Commit:'):
+                try:
+                    commit['committer'] = re.match(regex_committer, line).groupdict()
+                except AttributeError as e:
+                    if str(e).endswith("'groupdict'"):
+                        raise RuntimeError(f"Unable to parse committer-line:\n{line}") from e
+                    raise
+                continue
+            elif line.startswith('CommitDate:'):
+                commit['time'] = datetime.datetime.fromisoformat(line.split(maxsplit=1)[1])
+                continue
+            else:
+                # This line is part of the message.
+                if 'message' not in commit:
+                    commit['message'] = [line]
+                else:
+                    commit['message'].append(line)
+
+        return commit
+
+    def follow(self, path: Path | None = None, n: int | None = None) -> Generator[dict, None, None]:
+        """Yields commit dicts representing the history of ``path``.
+
+        History according to ``git log --follow`` if a path is given
+        or ``git log`` otherwise.
+
+        Parameters
+        ----------
+        path: Path, optional
+          What file to follow. If `None`, get the history of HEAD instead.
+        n: int, optional
+          Limit history going back ``n`` commits.
+          ``None`` for no limit (default).
+        """
+        # TODO: Do we want to distinguish methods for `git log` (path optional) vs `git log --follow` (requires a path)?
+        limit = [f'-n{n}'] if n is not None else []
+        pathspec = ['--follow', '--', str(path)] if path else []
+        cmd = ['log', '--date=iso-strict', '--pretty=fuller'] + limit + pathspec
+        output = self._git(cmd)
+
+        # yield output on a per-commit-basis
```

Comment:
That could work indeed.
I'll consider this in combination with your "whole string" suggestion (not line-splitting).
Kinda belongs together. Not sure about that, though.
For example: Things like stripping leading white spaces per line in the message may actually be more expensive via regex and while it's less code it may be harder to see what a given line of code is actually doing.

I'll try what that would look like. I do lean towards postponing that part, though. Need to implement the actual pseudo-keys and test they work and are queryable. And then there's the refactoring to be done, fully utilizing `Item` (and pass that around rather than `Path`s). This seems more urgent to get straight.